File: 2018/18/solution.py
def get_next(grid, x, y):
    num_trees = 0
    num_lumberyards = 0
    num_open = 0
    for dy in xrange(-1, 2):
        if y + dy < 0 or y + dy >= len(grid):
            continue
        for dx in xrange(-1, 2):
            if x + dx < 0 or x + dx >= len(max(grid)):
                continue
            n = grid[y + dy][x + dx]
            if n == '.':
                num_open += 1
            elif n == '|':
                num_trees += 1
            elif n == '#':
                num_lumberyards += 1
    cur = grid[y][x]
    if cur == '.' and num_trees >= 3:
        return '|'
    elif cur == '|' and num_lumberyards >= 3:
        return '#'
    elif cur == '#':
        if num_lumberyards >= 2 and num_trees >= 1:
            return '#'
        else:
            return '.'
    else:
        return cur


def get_next_grid(grid):
    new_grid = [['' for x in xrange(len(max(grid)))] for y in xrange(len(grid))]
    for y in xrange(len(grid)):
        for x in xrange(len(max(grid))):
            new_grid[y][x] = get_next(grid, x, y)
    return new_grid


def print_grid(grid):
    for y in xrange(len(grid)):
        string = ""
        for x in xrange(len(max(grid))):
            string += grid[y][x]
        print string


grid = []
with open('input') as f:
    for line in f:
        line = line.rstrip()
        grid.append(line)

flattened_grid = [item for sublist in grid for item in sublist]
flattened_str = ''.join(flattened_grid)
wooded = len([x for x in flattened_grid if x == '|'])
lumberyards = len([x for x in flattened_grid if x == '#'])

cache = []
for i in xrange(1, 500):
    if i % 10 == 0:
        print(i)
    grid = get_next_grid(grid)
    flattened = ''.join(''.join(row) for row in grid)
    if flattened in cache:
        cache_idx = cache.index(flattened) + 1 # list index is 0-based
        print("Found cycle, current iteration {} is repeat of {}".format(i, cache_idx))
        period = i - cache_idx
        while cache_idx % period != 1000000000 % period:
            cache_idx += 1
        print("Index {} has same remainder ({}) as 1000000000: {}".format(cache_idx, cache_idx % period, 1000000000 % period))
        wooded = len([x for x in cache[cache_idx-1] if x == '|'])
        lumberyards = len([x for x in cache[cache_idx-1] if x == '#'])
        print("Part 2: {}*{} = {}".format(wooded, lumberyards, wooded * lumberyards))
        break
    cache.append(flattened)

    if i == 10:
        wooded = len([x for x in flattened if x == '|'])
        lumberyards = len([x for x in flattened if x == '#'])
        print("Part 1: {}*{} = {}".format(wooded, lumberyards, wooded * lumberyards))

File: 2018/18/solution.cpp
#include <iostream>
#include <map>
#include <vector>

char get_next(std::vector<std::vector<char> >& grid, uint8_t x, uint8_t y) {
    uint8_t nTrees = 0;
    uint8_t nYards = 0;
    for ( int8_t dy = -1; dy <= 1; ++dy ) {
        if ( y + dy < 0 || y + dy >= grid.size() ) {
            continue;
        }
        for ( int8_t dx = -1; dx <= 1; ++dx ) {
            if ( x + dx < 0 || x + dx >= grid[0].size() ) {
                continue;
            }

            char n = grid[y + dy][x + dx];
            if ( n == '|' ) {
                nTrees++;
            } else if ( n == '#' ) {
                nYards++;
            }
        }
    }

    char cur = grid[y][x];
    if ( cur == '.' ) {
        return nTrees >= 3 ? '|' : '.';
    } else if ( cur == '|' ) {
        return nYards >= 3 ? '#' : '|';
    } else { // must be #
        if ( nYards >= 2 && nTrees >= 1 ) {
            return '#';
        } else {
            return '.';
        }
    }
}

std::vector<std::vector<char> > get_next_grid(std::vector<std::vector<char> >& grid) {
    std::vector<std::vector<char> > new_grid(grid.size(), std::vector<char>(grid[0].size(), ' '));
    for ( uint8_t y = 0; y < grid.size(); ++y ) {
        for ( uint8_t x = 0; x < grid[0].size(); ++x ) {
            new_grid[y][x] = get_next(grid, x, y);
        }
    }

    return new_grid;
}

int main() {
    std::vector<std::vector<char> > grid;
    while ( true ) {
        std::string s;
        std::cin >> s;
        if ( s.size() == 0 ) {
            break;
        }
        std::vector<char> row;
        for (auto& c : s) {
            row.push_back(c);
        }
        grid.push_back(row);
    }

    uint8_t nRows = grid.size();
    uint8_t nCols = grid[0].size();

    std::map<std::vector<std::vector<char> >, uint32_t> cache;
    uint32_t end = 1000000000;
    for ( uint32_t i = 1; i <= end; ++i ) {
        grid = get_next_grid(grid);
        if ( cache.count(grid) == 1 ) {
            // cycle!
            uint32_t add = (end - i) / (i - cache[grid]);
            i += add * (i - cache[grid]);
        } else {
            cache[grid] = i;
        }
        if ( i == 10 ) {
            uint32_t trees = 0;
            uint32_t yards = 0;
            for ( uint8_t r = 0; r < nRows; ++r ) {
                for ( uint8_t c = 0; c < nRows; ++c ) {
                    if ( grid[r][c] == '|' ) {
                        trees++;
                    } else if ( grid[r][c] == '#' ) {
                        yards++;
                    }
                }
            }
            std::cerr << "part 1: " << trees << " " << yards << " " << trees * yards << std::endl;
        } else if ( i == end ) {
            uint32_t trees = 0;
            uint32_t yards = 0;
            for ( uint8_t r = 0; r < nRows; ++r ) {
                for ( uint8_t c = 0; c < nRows; ++c ) {
                    if ( grid[r][c] == '|' ) {
                        trees++;
                    } else if ( grid[r][c] == '#' ) {
                        yards++;
                    }
                }
            }
            std::cerr << "part 2: " << trees << " " << yards << " " << trees * yards << std::endl;
        }
    }
}
